### PR TITLE
Improve handling of dynamic ELF symbols

### PR DESCRIPTION
**Pull request by d-e-s-o**
Please see individual commits for descriptions.

### CODE REVIEW

File: src/elf/parser.rs [hunk 1]
```diff
@@ -609,6 +663,26 @@ impl ElfParser {
         Ok(r)
     }
 
+    /// Perform an operation on each symbol.
+    pub(crate) fn for_each_sym<F, R>(&self, opts: &FindAddrOpts, r: R, mut f: F) -> Result<R>
+    where
+        F: FnMut(R, &SymInfo<'_>) -> R,
+    {
+        if let SymType::Variable = opts.sym_type {
```
Comment (anakryiko) [inline src/elf/parser.rs:684]: is SymType::Unknown expected and supported? If not, would it be better to make sure that opts.sym_type is exactly SymType::Function (for now?)?

**Review by anakryiko — APPROVED** [review]
Makes sense, LGTM.

File: src/elf/parser.rs (hunk 1 above)
Comment (danielocfb) [reply src/elf/parser.rs:684]: It's mirroring the existing logic we have in `find_addr`. If I were to guess, the reasoning was that `Unknown` basically represents "anything you have", and the data reported may change once we expose variables. Either way seems fine to me. Let me know what you prefer.

Comment (anakryiko) [reply src/elf/parser.rs:684]: ok, if Unknown means "don't care" then I guess this way is better.

**Comment by danielocfb** [comment]
Superseded by https://github.com/libbpf/blazesym/pull/472.

**Pull request closed without merging**